src/jira/util/jwt.ts

```diff
@@ -57,7 +74,27 @@ function sendError(res, code, msg) {
 	});
 }
 
-export const hasValidJwt = (secret: string, baseUrl: string, req: Request, res: Response) => {
+function verifyQsh(qsh: string, baseUrl: string, req: Request, res: Response) {
+	let expectedHash = jwt.createQueryStringHash(req, false, baseUrl);
+	let signatureHashVerified = qsh === expectedHash;
+	if (!signatureHashVerified) {
+		jwt.createCanonicalRequest(req, false, baseUrl); // eslint-disable-line
+
+		// If that didn't verify, it might be a post/put - check the request body too
+		expectedHash = jwt.createQueryStringHash(req, true, baseUrl);
+		signatureHashVerified = qsh === expectedHash;
+		if (!signatureHashVerified) {
+			jwt.createCanonicalRequest(req, true, baseUrl); // eslint-disable-line
+
+			// Send the error message for the first verification - it's 90% more likely to be the one we want.
+			sendError(res, 401, "Authentication failed: query hash does not match.");
+			return false;
+		}
+	}
+	return true
+}
+
+export const hasValidJwt = (secret: string, baseUrl: string, req: Request, res: Response, tokenType: TokenType) => {
```

Comment:
I'm really not a fan of passing res/req to internal function as it creates spaghetti code and you lose who's in control of what.

All things related to the web server should be handled by the route controller to prevent stepping on toes and reducing complexity.

Reply:
I am a bit reluctant to refactor this part here, hence I would like to minimise the changes. However I can use it as a deleage of middleware function, but adding "next" parameter and calling it from there. Then we can rename it to "validateJwtMiddleware"